roles/jd-server/src/lib/job_declarator/message_handler.rs

```diff
@@ -38,6 +41,53 @@ impl JobDeclaratorDownstream {
     }
 }
 
+pub fn clear_declared_mining_job(
+    mining_job: DeclareMiningJob,
+    mempool: Arc<Mutex<JDsMempool>>,
+) -> Result<(), Error> {
+    // If there is an old declared mining job, remove its transactions from the mempool
+    // Retrieve necessary data from the old job
+    let transactions_to_remove = mining_job.tx_short_hash_list.inner_as_ref();
+    if transactions_to_remove.is_empty() {
+        info!("No transactions to remove from mempool");
+        return Ok(());
+    }
+
+    let nonce = mining_job.tx_short_hash_nonce;
+
+    for short_id in transactions_to_remove {
+        let result = mempool.safe_lock(|mempool_| -> Result<(), Error> {
+            // Try to manage this unwrap, we use .ok_or() method to return the proper error
+            let short_ids_map = mempool_
```

Comment:
do we need to call `mempool.safe_lock` in each iteration? couldnt we retrieve `short_ids_map` before the `for` and just use it?

Reply:
This was a feedback from lorban I implemented and modified. If you have a better approach I can compare that

Reply:
That means I would still need to handle the double unwraps well? @lorbax

Reply:
yes, but I suggest first to write a the code as easy as possible, thest it, then add error handling and then test it deeply. You should check that the code you introduced does what it is supposed to do. In particular, the old job must be removed from jds mempool and while updating the jds mempool the fat transactions are not discarded.
You can do that using testnet4, or a custom signet

Reply:
I'm worried that to_short_ids is very expensive not sure if is really better to lock the mutex for all that time. Btw we should have a benchmark for cases where there is contention for the lock.

Reply:
considering that this is JDC and not JDS it shouldn't be an issue as we expect to have only one connection

Reply:
this should look into the new received DeclareMiningJob and remove only the txs that are in the old one but not in the new one. Also better would be to have a rc for each txs and decrement it everytime that the tx is in the old one and not in the new one (this cause we have multiple downstreams). And when we reach 0 we remove the tx

Reply:
somthing like that:  `    pub mempool: HashMap<Txid, Option<(Transaction,u8)>>`

Reply:
> this should look into the new received DeclareMiningJob and remove only the txs that are in the old one but not in the new one. Also better would be to have a rc for each txs and decrement it everytime that the tx is in the old one and not in the new one (this cause we have multiple downstreams). And when we reach 0 we remove the tx

I implemented this suggestion here https://github.com/plebhash/stratum/commit/24b952a296b2949729e85ea4b29e5e58b356824b

Reply:
> somthing like that: ` pub mempool: HashMap<Txid, Option<(Transaction,u64)>>`

Could we do take advantage of the `token` or other property in `JobDeclaratorDownstream` in order to do that?